### ISSUE TITLE
function_score's min_score option does not play well with caching

### ISSUE THREAD

**Issue by jpountz** [elastic/elasticsearch#14300]
`function_score` has a `min_score` option that allows to reject documents whose score is not above a given threshold. The problem is that the produced scores depend on the query norm/boost so scores might be different depending on where the `function_score` query is in the query tree, which makes it impossible to cache.

Here are the options I can think of:
 - only allow to use `min_score` in combination with `boost_mode=replace` (which does not take normalization factors into account)
 - stop supporting `min_score`
 - add a hack to disable caching when `min_score` is used

**Comment by clintongormley** [comment]
@jpountz just to give some context: imagine your query is a has_child query which is using this technique to score parents https://github.com/elastic/elasticsearch/issues/2917, you can use `min_score` to exclude parents that don't have a high enough score.

**Comment by jpountz** [comment]
Right, I understand that this can be useful. I'm just concerned that if you use this feature in a sub-tree of the query that doesn't need scores (ie. either a filter or the top-level query if you don't sort on the score) then you will hit bugs.

**Comment by clintongormley** [comment]
With the changes of boost -> boost query and the switch in Lucene 6 to BM25, which doesn't use query norm, would it be possible to cache this query in those circumstances?
